### DISCUSSION TITLE
Snap endpoints

### DISCUSSION THREAD

**Discussion by JimMadge** [alan-turing-institute/data-safe-haven#1252] [Ideas]
## :computer: System information

- **Data Safe Haven version:** 3.4.0+

## :no_entry_sign: Problematic behaviour

The SRD includes some snap packages. Snaps are automatically updated by the snap daemon. Some set of endpoints will need to be reachable for this to work.

## :steam_locomotive: Possible workarounds, remediations or solutions

[This](https://askubuntu.com/questions/1141546/what-firewall-port-is-required-for-snap-package-management) stackoverflow post contains a suggested list of urls.

From [The snapcraft website](https://snapcraft.io/docs/network-requirements#heading--api),

> Download CDNs
>
>    *.snapcraftcontent.com:443
>    *.cdn.snapcraft.io:443 [deprecated]
>    public.apps.ubuntu.com:443 [deprecated]

Hopefully this is all that is required. The Snapcraft API may not be needed if you don't intend to publish packages (which we would not want to allow).

**Comment by JimMadge** [comment]
[This](https://snapcraft.io/docs/network-requirements) looks to be the authoritative answer.

Note that,

snap / snap daemon is the client that downloads/installs/managed packages
snapcraft is the tool for *building* and *publishing* snaps

**Comment by jemrobinson** [comment]
Can you get a list of URLs and IP addresses and add it to the top-level comment on this issue?

**Comment by JimMadge** [comment]
Are both IPs and URLs needed?

**Comment by jemrobinson** [comment]
Yes, we use IPs to restrict outbound access at the NSG level and then FQDNs at the Firewall.

**Comment by JimMadge** [comment]
Ok, I can't find any indication that the IP addresses are guaranteed to be stable.

**Comment by jemrobinson** [comment]
They don't have to be completely fixed, we just need to know the range of IP addresses that they are drawn from. For example if they use a CDN then all IP addresses that the CDN might resolve to.

**Comment by jemrobinson** [comment]
We also get a lot of outbound requests to `api.snapcraft.io` that are denied by the firewall. Do we want to allow these?

**Comment by jemrobinson** [comment]
Looks like there are some fixed domains [here](https://forum.snapcraft.io/t/network-requirements/5147)

- storage.snapcraftcontent.com:443
- canonical-lgw01.cdn.snapcraftcontent.com:443
- canonical-lcy01.cdn.snapcraftcontent.com:443
- canonical-lcy02.cdn.snapcraftcontent.com:443
- canonical-bos01.cdn.snapcraftcontent.com:443
- cloudfront.cdn.snapcraftcontent.com:443 [used inside AWS]
- fastly.cdn.snapcraftcontent.com:443 [deprecated]
- fastly-global.cdn.snapcraftcontent.com:443 [deprecated]

**Comment by JimMadge** [comment]
I'm uncertain about `api.snapcraft.io`. The network requirements page lists that as one of three endpoints required for API access. This _might_ include publishing packages so could be an egress risk. However, it isn't clear. It could just be `snapd` using the API to look for new versions of installed snaps.

**Comment by jemrobinson** [comment]
I think the current requests are coming from an update checker (presumably `snapd`). The question is: if we allow this, will it represent an egress risk (and similarly, if we **don't** allow this, will it break snap updates)?

**Comment by JimMadge** [comment]
> Snapcraft also requires:
>
>    [upload.apps.ubuntu.com:443](http://upload.apps.ubuntu.com:443/)

Which sounds like an endpoint for publishing, and is only needed for snapcraft. But the docs are not explicit about what these endpoints are for.

> The question is: if we allow this, will it represent an egress risk (and similarly, if we don't allow this, will it break snap updates)?

The second of those is easy to test with an Ubuntu VM and some `ufw` commands. The first isn't.

**Comment by JimMadge** [comment]
[Snap proxy](https://docs.ubuntu.com/snap-store-proxy/en/) (_maybe_ something for the future).
